### ISSUE TITLE
kube-DNS request/limits messes with tail latency

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#33027]
Long tail observed in normal case: 

``` console
$ for i in `seq 1 1000`; do dig @10.0.0.10
kubernetes.svc.cluster.local A | grep Query | cut -f2 -d: ; done |
sort -n | uniq -c

      4  0 msec
    930  1 msec
     29  2 msec
     11  3 msec
      1  4 msec
      4  5 msec
      1  6 msec
      2  7 msec
      1  9 msec
      1  10 msec
      1  13 msec
      1  15 msec
      1  22 msec
      1  29 msec
      2  44 msec
      1  48 msec
      1  51 msec
      1  65 msec
      1  68 msec
      2  72 msec
      2  73 msec
      1  76 msec
      1  83 msec
```

I tried a bunch of things like using hostNetwork and doing a lookup from 127.0.0.1, didn't help.

Then tried:
- drained the node (deleted all pods, cordoned the node) 
- removed dnsmasq from kube-dns yaml
- removed static pod manifests for fluentd, kube-proxy from /etc/kubernetes
- modified podTemplate spec like of kube-dns rc to 

``` yaml
    spec:
      containers:
      - args:
        - --domain=cluster.local.
        - --dns-port=53
        image: gcr.io/google_containers/kubedns-amd64:1.7
        imagePullPolicy: IfNotPresent
        name: kubedns
        ports:
        - containerPort: 53
          hostPort: 53
          name: dns-local
          protocol: UDP
        resources:
          requests:
            cpu: 200m
            memory: 70Mi
        terminationMessagePath: /dev/termination-log
      dnsPolicy: ClusterFirst
      nodeName: e2e-test-beeps-minion-group-oka0
      restartPolicy: Always
      securityContext: {}
      terminationGracePeriodSeconds: 30
```

basically kube-dns now has 1 box.

Observed much better latencies, at least in the handful of lookups I did: 

```
root@dnsutils2:/# for i in `seq 1 1000`; do dig @10.0.0.10 kubernetes.svc.cluster.local A | grep Query | cut -f2 -d: ; done | sort -n | uniq -c
    537  0 msec
    456  1 msec
      4  2 msec
      2  3 msec
      1  9 msec
```

Theories:
1. dnsmasq is cpu constrainted (it has 10m)
2. kube-dns is setting GOMAXPROCS to what it _thinks_ is num cores
3. one of the other components I removed from the node was hammering cpu 

We should try doing https://github.com/kubernetes/kubernetes/issues/29597#issuecomment-235469479

**Comment by bprashanth** [comment]
This *may* have been fixed when we lifted the limits on dnsmasq, but IIUC the previous repro was on container-vm, which never enforced limits anyway. Worth trying to stress the node because then dnsmasq only gets its request.

**Comment by bowei** [comment]
# DNS latency

## TLDR

It looks like the tail latency issue is no longer an issue. I suggest increasing the CPU allocation to 200mi for safety (e.g. 20% of one core).

## Details

Using the script snippet from beeps@ and varying the amount of CPU utilization of the node (excluding from other system services and kube-dns) that the DNS
server is scheduled on:

cpu usage %|50%tile (ms)| 90|95|99|99.5|
--------- | ---- | ---- | ---- | ---- | ---- |
0         | 0    | 3    | 9    | 19   | 23   |
50        | 0    | 1    | 4    | 11   | 13   |
80        | 0    | 2    | 8    | 17   | 20   |
100       | 0    | 3    | 11   | 21   | 24   |
120       | 0    | 7    | 15   | 31   | 36   |
140       | 0    | 15   | 23   | 37   | 44   |

Note that full utilization is 200% as this is a 2-vCPU machine. I wasn't able to schedule any more pods onto the node as it was full (rest was reserved for
kube-dns and system pods).

We see that the tail latency for the higher percentiles remain  < 50ms.

## Notes

* rc-xx.txt contains a table of (latency, count) for xx*10 cpu utilization. CPU
  was allocated in 10% chunks at a time, xx represents the number of chunks
  scheduled.

### Latency script

````
#!/bin/bash

echo > /tmp/out
for i in `seq 1 1000`; do
        dig @10.0.0.10 kubernetes.svc.cluster.local A \
                | tee -a /tmp/out \
                >/dev/null
        echo -n .
done
cat /tmp/out | grep Query | cut -f2 -d: | sort -n | uniq -c
````

### Replication controller yaml for cpu-hog 
 
```` 
apiVersion: v1 
kind: ReplicationController 
metadata: 
  creationTimestamp: 2016-10-24T23:59:23Z 
  generation: 17 
  labels: 
    app: bowei-tk 
  name: bowei-tk 
  namespace: default 
  resourceVersion: "542866" 
  selfLink: /api/v1/namespaces/default/replicationcontrollers/bowei-tk 
  uid: e340eb27-9a45-11e6-ab47-42010af00002 
spec: 
  replicas: 8 
  selector: 
    app: bowei-tk 
  template: 
    metadata: 
      creationTimestamp: null 
      labels: 
        app: bowei-tk 
      name: bowei-tk 
    spec: 
      containers: 
      - image: gcr.io/bowei-gke-dev/bowei-tk-cpu-hog-amd64 
        imagePullPolicy: Always 
        name: bowei-tk 
        ports: 
        - containerPort: 80 
          protocol: TCP 
        resources: 
          limits: 
            cpu: 100m 
            memory: 170Mi 
          requests: 
            cpu: 100m 
            memory: 10Mi 
        terminationMessagePath: /dev/termination-log 
      dnsPolicy: ClusterFirst 
      restartPolicy: Always 
      securityContext: {} 
      terminationGracePeriodSeconds: 30 
````

**Comment by bowei** [comment]
[dns-latency.zip](https://github.com/kubernetes/kubernetes/files/551921/dns-latency.zip)

**Comment by bowei** [comment]
we have adjusted the resource limits for the dns pods, this should be less of a problem
